############################################## Tajimas D ##############################################

#!/bin/bash
#$ -pe smp 6
#$ -l h_vmem=16G
#$ -l h_rt=72:0:0
#$ -cwd
#$ -j y
#$ -N  filter_stats
#$ -o /data/scratch/bt24076
#$ -m bea
#$ -l rocky

set -e 

#load modules
module load bcftools
module load vcftools

#loop for all chromsomes
for i in {1..22} X;
do

path to vcfs
path_to_vcf="/data/scratch/bt24076/03-02-2025-T2D_pjct_Data/ALL.chr${i}.phase3_shapeit2_mvncall_integrated_v5b.20130502.genotypes.vcf.gz"

bangladesh
bcftools view --threads $NSLOTS -S /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/benbang.txt -o /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/beb_filt/benbang_CHR${i}_filt.vcf $path_to_vcf

indian in hou
bcftools view --threads $NSLOTS -S /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/gujIndHou.txt -o /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/gih_filt/gujIndHou_CHR${i}_filt.vcf $path_to_vcf

pakistan
bcftools view --threads $NSLOTS -S /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/pjlPak.txt -o /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/pjl_filt/pjlPak_CHR${i}_filt.vcf $path_to_vcf

indian Tel UK
bcftools view --threads $NSLOTS -S /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/indTelUK.txt -o /data/scratch/bt24076/03-02-2025-T2D_pjct_Data/itu_filt/indTelUK_CHR${i}_filt.vcf $path_to_vcf
done


#Tajim's D on all Chrom and Populations
for i in {1..22} X;
do

#paths to filtered south asain popultaions
beb_vcf="/data/scratch/bt24076/03-02-2025-T2D_pjct_Data/beb_filt/benbang_CHR${i}_filt.vcf"
gih_vcf="/data/scratch/bt24076/03-02-2025-T2D_pjct_Data/gih_filt/gujIndHou_CHR${i}_filt.vcf"
itu_vcf="/data/scratch/bt24076/03-02-2025-T2D_pjct_Data/itu_filt/indTelUK_CHR${i}_filt.vcf"
pjl_vcf="/data/scratch/bt24076/03-02-2025-T2D_pjct_Data/pjl_filt/pjlPak_CHR${i}_filt.vcf"

#Tajimas D calculation
vcftools --vcf $beb_vcf --TajimaD 10000 --out /data/scratch/bt24076/11-02-2025_psResults/beb/beb_CHR${i}

vcftools --vcf $gih_vcf --TajimaD 10000 --out /data/scratch/bt24076/11-02-2025_psResults/gih/gih_CHR${i}

vcftools --vcf $itu_vcf --TajimaD 10000 --out /data/scratch/bt24076/11-02-2025_psResults/itu/itu_CHR${i}

vcftools --vcf $pjl_vcf --TajimaD 10000 --out /data/scratch/bt24076/11-02-2025_psResults/pjl/pjl_CHR${i}

done
############################################## nSL ##############################################

# Manually caclulated nSL from data retrieved from filtered vcf files which includes subpopulations
# Please Edit Accordingly

#!/bin/bash
#$ -pe smp 4
#$ -l h_rt=72:0:0
#$ -l h_vmem=6G
# Request 10 CPU cores
# Request 72-hour runtime (i.e., 3 days)
# Request 6GB RAM per core (i.e., 60GB total)
#$ -cwd
#$ -j y
#$ -m beas
#$ -l rocky
set -e
# Generated by Job Script Builder on 2025-02-11
# For assistance, please email its-research-support@qmul.ac.uk
# Please cite use of Apocrita in your Research
# See https://docs.hpc.qmul.ac.uk/using/citing/ for details     

SELSCAN_PATH="/data/scratch/bt24041/selscan/bin/linux/norm"

# Run SELSCAN with the --out flag to specify the output file
$SELSCAN_PATH --nsl --vcf /data/scratch/bt24041/input/output/biallelic_only{chromosomeNumber}.vcf --out /data/scratch/bt24041/output/{subpopulation}{chromosomeNumber}_nsl

echo "Selscan analysis complete. Output saved to /data/scratch/bt24041/output."
~                                                                                                                                                                                                           
~                              


~